frontend/src/metabase/common/components/EntityPicker/components/NestedItemPicker/NestedItemPicker.tsx

```diff
@@ -0,0 +1,87 @@
+import type React from "react";
+
+import ErrorBoundary from "metabase/ErrorBoundary";
+import { Flex } from "metabase/ui";
+import type { SearchListQuery } from "metabase-types/api";
+
+import type {
+  EntityPickerOptions,
+  TypeWithModel,
+  PickerState,
+  TisFolder,
+} from "../../types";
+import type { EntityItemListProps } from "../ItemList";
+
+import { AutoScrollBox } from "./AutoScrollBox";
+import { ListBox } from "./NestedItemPicker.styled";
+
+export interface NestedItemPickerProps<TItem extends TypeWithModel> {
+  onFolderSelect: ({ folder }: { folder: TItem }) => void;
+  onItemSelect: (item: TItem) => void;
+  itemName: string;
+  options: EntityPickerOptions;
+  path: PickerState<TItem>;
+  isFolder: TisFolder<TItem>;
+  listResolver: React.FC<
+    EntityItemListProps<TItem> & {
+      options: EntityPickerOptions;
+    }
+  >;
+}
+
+const generateKey = (query?: SearchListQuery) =>
+  JSON.stringify(query ?? "root").slice(0, 255);
+
+export function NestedItemPicker<TItem extends TypeWithModel>({
+  onFolderSelect,
+  onItemSelect,
+  itemName,
+  options,
+  path,
+  isFolder,
+  listResolver: ListResolver,
+}: NestedItemPickerProps<TItem>) {
+  const handleFolderSelect = (folder: TItem) => {
+    onFolderSelect({ folder });
+  };
+
+  const handleClick = (item: TItem) => {
+    if (isFolder(item)) {
+      handleFolderSelect(item);
+    } else {
+      onItemSelect(item);
+    }
+  };
+
+  return (
+    <AutoScrollBox
+      data-testid="nested-item-picker"
+      contentHash={generateKey(path[path.length - 1].query)}
```

Comment:
```suggestion
      contentHash={generateKey(_.last(path)?.query)}
```
index access is unsafe